### PR TITLE
Don't add column for SSB with only one step type

### LINKED ISSUE #95: SimpleStepSelectorBuilder case when there is only one step
The SimpleStepSelectorBuilder creates a column for each step type, that gets to be 1 in the row where a step instance of that type starts.

When there is only one step type we don't need any column, because all steps are of this step type.

Make SimpleStepSelectorBuilder to don't add any columns where there is a column, and create a StepSelector that works without any column.

### PR DESCRIPTION
**Pull request by 000wan**
# Summary
It resolves #95.

It makes SimpleStepSelectorBuilder not add any step selector columns if there is only one step.
Further, I wrote a unit test for this (SSB with single step) and it might be helpful for #102.

# Test Results
Passes all tests with `cargo test`.
Runs Ok with `/examples/fibonacci.rs`, `/examples/mimc7.rs`, `/examples/poseidon.rs`
Witness display results of `cargo run --package chiquito --example fibonacci` changed:

### Before:
```
offset,srcm forward a,srcm forward b,srcm internal signal c,'step selector for fibo step'
0,1,1,2,1
1,1,2,3,1
2,2,3,5,1
3,3,5,8,1
4,5,8,13,1
5,8,13,21,1
6,13,21,34,1
7,21,34,55,1
8,34,55,89,1
9,55,89,144,1
10,89,144,233,1
```
### After:
```
offset,srcm forward a,srcm forward b,srcm internal signal c
0,1,1,2
1,1,2,3
2,2,3,5
3,3,5,8
4,5,8,13
5,8,13,21
6,13,21,34
7,21,34,55
8,34,55,89
9,55,89,144
10,89,144,233
```

### CODE REVIEW

**Comment by 000wan** [comment]
Looking forward to your reviews and comments!
Btw, I'm currently into this SSB code so it would be wonderful if there are other issues/features related to this that I can work on.

File: src/plonkish/compiler/step_selector.rs [hunk 1]
```diff
@@ -74,6 +74,26 @@ impl StepSelectorBuilder for SimpleStepSelectorBuilder {
...
+                .selector_expr
+                .insert(step.uuid(), PolyExpr::Const(F::ONE));
+
+            selector
+                .selector_expr_not
+                .insert(step.uuid(), PolyExpr::Const(F::ZERO));
+
+            selector
+                .selector_assignment
+                .insert(step.uuid(), vec![(PolyExpr::Const(F::ONE), F::ONE)]);
```
Comment (leolara) [inline src/plonkish/compiler/step_selector.rs:91]: @000wan this looks not the more readable way. I would make get_selector_assignment to return an Option, and then the code in assignment generator, could ignore this if None is returned.

Comment (000wan) [reply src/plonkish/compiler/step_selector.rs]: I agree with that. Now I modified `selector_assignment` to have an Option value. Could you check this again? @leolara

File: src/plonkish/compiler/step_selector.rs [hunk 2]
```diff
@@ -16,7 +16,7 @@ pub type SelectorAssignment<F> = (PolyExpr<F>, F);
 pub struct StepSelector<F> {
     pub selector_expr: HashMap<StepTypeUUID, PolyExpr<F>>,
     pub selector_expr_not: HashMap<StepTypeUUID, PolyExpr<F>>,
-    pub selector_assignment: HashMap<StepTypeUUID, Vec<SelectorAssignment<F>>>,
+    pub selector_assignment: HashMap<StepTypeUUID, Option<Vec<SelectorAssignment<F>>>>,
```
Comment (000wan) [inline src/plonkish/compiler/step_selector.rs:19]: I made this Option in order to distinguish the cases where the uuid is invalid, or the uuid is valid but has no assignments.

**Comment by leolara** [comment]
@000wan we are in a hacker house mentoring, that is why we have not answered in so long. This PR will get merged for sure. Let me get back to you

**Comment by 000wan** [comment]
I didn't know you were busy.. then I'll wait for it

**Comment by leolara** [comment]
@000wan so the only problem now is that if the step has more than one row, then it will be active also for the rows that is not the first row of the step. The best way to solve this is to make `q_enable` to be only one at the beginning of each step. this change of course, only if the number of steps is one.

So currently q_enable is 1 for all rows. if we have only one step with row height 3, then q_enable should be 1,0,0,1,0,0,1,0,0,.. and so on

**Comment by 000wan** [comment]
Then doesn't it require a new column for the step?

**Pull request closed without merging**